Cy8cproto_041tp TCPWM_COUNTER Driver

Deepika-aerlync opened the pull request:
This Pull Request depends on : https://github.com/zephyrproject-rtos/zephyr/pull/100894

JarmouniA commented on include/zephyr/drivers/timer/ifx_tcpwm.h:
Why is this header in `Zephyr/include`?

Deepika-aerlync replied:
Removed.

teburd replied:
Will take a look, thanks for pointing this out. You are absolutely right this shouldn't probably be in the public include directory, though it looks like we aren't the only ones doing so...

Deepika-aerlync committed:
drivers: counter: Add Infineon counter driver

Add implementation of the Infineon PSoC4 TCPWM-based counter driver.
- Provides basic counter operations
including start, stop, read, and set alarm.
- Supports configuration and initialization through Device Tree.
- Enables alarm callback handling for
precise time-based event generation.

Signed-off-by: Deepika R <deepika@aerlync.com>
Signed-off-by: Sayooj K Karun <sayooj@aerlync.com>

Deepika-aerlync committed:
samples: drivers: counter: alarm: boards: Add overlay file

- Add an overlay to enable TCPWM Counter sample.
- Contains configurations for Counter and Clock

Signed-off-by: Deepika R <deepika@aerlync.com>
Signed-off-by: Sayooj K Karun <sayooj@aerlync.com>

Deepika-aerlync committed:
dts: arm: infineon: Change TCPWM Base Register

-Change base register for TCPWM to 0x40200000
 which controls the entire IP block.
- enabling/disabling counters and handling
 global interrupts.

Signed-off-by: Deepika R <deepika@aerlync.com>
Signed-off-by: Sayooj K Karun <sayooj@aerlync.com>

Deepika-aerlync commented on dts/arm/infineon/psoc4/psoc4100tp/psoc4100tp.dtsi:
```diff
@@ -23,18 +23,18 @@
 
 	soc {
 		pinctrl: pinctrl@40020000 {
-			compatible = "infineon,cat1-pinctrl";
```
Fixed.

sreeramIfx commented on drivers/counter/counter_infineon_tcpwm.c:
```diff
@@ -459,23 +486,29 @@ static DEVICE_API(counter, counter_api) = {
+#else
+#define TCPWM_CNT_IDX(n)                                                                           \
+	.index = (DT_REG_ADDR(DT_INST_PARENT(n)) - DT_REG_ADDR(DT_PARENT(DT_INST_PARENT(n)))) /    \
+		 DT_REG_SIZE(DT_INST_PARENT(n))
```
My intention was that `DT_NODE_CHILD_IDX(DT_INST_PARENT(n))` can be used for other devices as well.

Deepika-aerlync replied:
```DT_NODE_CHILD_IDX(DT_INST_PARENT(n))``` even though we make it generic for other devices, our side we only able compile/build but not able to test this with other devices.

sreeramIfx commented on dts/arm/infineon/psoc4/psoc4100tp/psoc4100tp.dtsi:
```diff
@@ -128,8 +128,8 @@
-		tcpwm0: tcpwm0@40200100 {
-			reg = <0x40200100 0x180>;
+		tcpwm0: tcpwm0@40200000 {
+			reg = <0x40200000 0x300>;
```
Even 0x280 does not make sense
